File: first_last_elem.java
// Time Complexity : O(Log(N)) -> Binary search involved
// Space Complexity : O(1) -> No extra space used
// Did this code successfully run on Leetcode : YES
// Any problem you faced while coding this :
//Couldn't solve problem on my own.

// Your code here along with comments explaining your approach


class Solution {
    public int[] searchRange(int[] nums, int target) {
        
        if(nums == null || nums.length == 0)
            return new int[] {-1, -1};
        
        int firstElem = findRightMostIndex(nums, target-1);
        int lastElem = findRightMostIndex(nums, target); 
        
        if(firstElem == lastElem)
            return new int[]{-1, -1};
        
        return new int[] {firstElem + 1, lastElem};
        
    }
        
        private int findRightMostIndex(int[] arr, int value){
            
            int low = 0;
            int high = arr.length-1;            
            
            if(value < arr[0])
                return -1;
            
            if(value >= arr[arr.length-1])
                return arr.length-1;
            
            while(low<=high){
                int mid = low + (high-low)/2;
                
                if(arr[mid] <= value && arr[mid+1] > value){
                    return mid;
                }
                if(arr[mid] <= value){
                    low = mid + 1;
                }else{
                    high = mid - 1;
                }
            }
            
            return -1;
        }
}

File: firstLastElem.py
# Time Complexity : O(log n)
# Space Complexity : O(n)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this :


class Solution(object):
        def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if nums == []:
                return [-1,-1]
        if target not in nums:
                return [-1,-1]
        if len(nums) == 1:
                return [0,0]
        l=0
        r=len(nums)-1
        mid = 0
        while(l<=r):
                mid = l + (r-l)/2
                if nums[mid] == target:
                break
                elif nums[mid] > target:
                r = mid-1
                else:
                l = mid+1
        #print(mid)
        l = mid
        r = mid
        i=0
        j=0
        retlist = [0,0]
        for i in reversed(range(mid+1)):
                if nums[i] == target:
                        retlist[0] = i
        for j in range(mid, len(nums)):
                if nums[j] == target:
                        retlist[1] = j
        return retlist